### ISSUE TITLE
border_mode='same' and/or custom zero padding for convolutions

### ISSUE THREAD

**Issue by benanne** [Theano/Theano#2118]
Several convolution implementations (cuda-convnet, cudnn, ...) provide a way to implicitly pad the input with a border of zeros. This mechanism is then used to implement the various 'border modes' (i.e. full, valid, same).

Currently Theano's `theano.tensor.nnet.conv2d` only allows to set the `border_mode`, and as far as I can tell all current implementations only support `valid` and `full`, but not `same`.

The main advantage of `border_mode='same'` is that the output feature maps are the same size as the input, so this makes it a lot easier to design architectures with lots of stacked layers.

Implementing a `same` convolution in Theano is of course possible by doing the padding manually and then using a `valid` convolution, or by using a `full` convolution and then slicing the output. But neither of these methods is optimal.

Considering that `same` convolutions are pretty commonly used now in state of the art networks (Krizhevky's 2012 ImageNet entry, VGG, GoogLeNet, ...), it would be useful to have proper support for this in Theano as well.

The easiest way to achieve this would be to make `theano.tensor.nnet.conv2d` support `border_mode='same'`. The most flexible way would be to explicitly expose a `pad` argument, like cuda-convnet and cudnn. The most convenient way would be to support both :)

I don't really have any code to contribute unfortunately, but I wanted to bring this up in case someone else is interested in this and wants to work on it.

TODO:
- [x] custom padding gh-2259
- [x] border=half gh-2126
- [ ] border=same

**Comment by jwingit** [comment]
A simple way to emulate border_mode=half is to simply access a subset of the conv2D output. E.g., for square filter you could define  this module:

    def conv2D_border_mode_same( x, w ):
        fso = T.shape(w)[2] - 1  # this is the filter size minus 1
        return conv2d(x, wu, border_mode='full')[:,:,fso:nps+fso,fso:nps+fso]

**Comment by nouiz** [comment]
Yes, but that trigger extra computation. It would be better to not do that.
But you give a quick workaround. This is useful until it get done. There is
a PR that have a new CPU convolution with custom padding. We already have
that for cudnn.

**Comment by nouiz** [comment]
border mode is done for some time already. We probably won't implement border_mode=same, so closing.

**Comment by ragavvenkatesan** [comment]
`        if border_mode == 'same':               # this is used typically by VGG.
            _out_height = image_shape[2]
            _out_width = image_shape[3]
            x = (filter_shape[2] - 1) // 2
            y = (filter_shape[3] - 1) // 2
            border_mode = (x,y)

            from theano.sandbox import cuda
            if not cuda.dnn.dnn_available():
                raise Exception ("cuDNN is needed for this type of convolution.")

            else:
                if subsample[0] == 1 and subsample[1] == 1:
                    self.out = cuda.dnn.dnn_conv (
                            img = input ,
                            kerns = filters,
                            border_mode= border_mode,
                            conv_mode = 'cross'
                            )
                else:
                    self.out = cuda.dnn.dnn_conv (
                            img = input ,
                            kerns = filters,
                            border_mode= border_mode,
                            subsample = subsample,
                            conv_mode = 'cross'
                            )`

Here is how I do this at the moment, but with theano 0.10.0 cuda backend is going to removed. I wonder if there is a reason why same won't be supported in theano using the libgpuarray backed?

**Comment by nouiz** [comment]
same should work with the new back end. If not, make sure to update Theano
to version 0.9rc3 and update libgpuarray to 0.6.2. if the problem persist,
open a new issue with the full error message.

Le mar. 7 mars 2017 14:33, Ragav Venkatesan <notifications@github.com> a
écrit :

**Comment by f0k** [comment]
> Here is how I do this at the moment

Looks like you'd just want to use `border_mode='half'`: http://deeplearning.net/software/theano/library/tensor/nnet/conv.html#theano.tensor.nnet.conv2d
There is no implementation of `border_mode='same'` since that would require asymmetric padding for even filter sizes. For odd filter sizes, `border_mode='half'` does exactly what you want. Sorry if the naming is confusing, I guess that was based on my suggestion, to make clear that it does symmetric padding with half the filter size.

**Comment by rjpg** [comment]
In the "same" mode (zero padding) it would be nice to have something like "same donut" and "same cylinder" to choose "top/bottom"  of "side to side" and instead of zeros it would "roll" the extremes of the images ... For example the top zeros would be the values on the bottom of the image.    This would be nice in multivariables time-series problems , where I have better results than using LSTM ... (Instead of a CNN looking to 2D images it looks into a 2D map of several values through time ). The kernel map (CNN "neuron") instead of making correlations with "zeros", while working on the edge of the multivariable time-serie map, it would be nice to make correlations with with the variables from top to bottom... 

Well it is difficult to explain ... Where can I see the source code of keras tensorflow where they put zeros around the image ? so I can make a new version for me to pout other thins instead of zeros ?

**Comment by lamblin** [comment]
#5827 is about that feature request.
It is unlikely we will work on integrating this in the interface, but you could probably do something with subtensors and concatenation to emulate it.

**Comment by f0k** [comment]
> Where can I see the source code of keras tensorflow where they put zeros around the image ?

On CUDA GPUs, the actual implementation is usually in cuDNN, which is a closed-source library provided by Nvidia. Libraries will usually have fall-back implementations that you could modify (often based on im2col and gemm), but are a bit slower. The easiest workaround will be to pad your network input (with wraparound) and then use unpadded ("valid") convolutions throughout.

**Comment by rjpg** [comment]
yes nice ideia ... I will build some custom layer with using keras lambda function to transform the maps and include the lines on top and bottom of the feature with the values of the bottom and top lines of the feature to create an horizontal cylinder ... according with the size of the filter used in the next layer (half of it...). Nice Ideia ! thanks
